2290 Add support to update data in Datasets

Nic-Ma opened the pull request:
Fixes #2290 .

### Description
This PR added support to update data after several epochs in Datasets.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

Nic-Ma commented:
/black

Nic-Ma commented:
/black

wyli commented on monai/data/dataset.py:
```diff
@@ -173,6 +176,17 @@ def __init__(
             if not self.cache_dir.is_dir():
                 raise ValueError("cache_dir must be a directory.")
 
+    def update_data(self, data: Sequence):
```
LMDBDataset inherits this but probably wouldn't work

Nic-Ma replied:
Cool, thanks for your finding!
I added support to LMDBDataset in the latest commit and unit tests.

Thanks.

wyli reviewed (commented):
I think the new method could be `set_data` instead of `update_data`? not sure about the overall stability in the multiprocess loading because there will be concurrent read and write for the cache. maybe the PersistentDataset is ok but how about the other datasets?

Nic-Ma commented on monai/data/dataset.py:
```diff
@@ -173,6 +176,17 @@ def __init__(
+        self.data = data
+
+        if self.cache_dir is not None and self.cache_dir.exists():
+            shutil.rmtree(self.cache_dir)
```
updated in the latest commit.
Thanks.

Nic-Ma commented on monai/data/dataset.py:
```diff
@@ -711,6 +744,22 @@ def __init__(
+        if self.shuffle:
+            data = copy(data)
+            self.randomize(data)
+        return super().update_data(data)
```
Thanks for pointing out it, updated in the latest commit.

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @wyli ,

Thanks for your review and suggestions.
I updated the PR according to your comments.
Could you please help review it again?
I already added multi-processing test based on workers of DataLoader:
https://github.com/Project-MONAI/MONAI/pull/2436/files#diff-eea046ab7338e5acd9b172b24ec5a704a755dd10bfae2056b046d598d064ecf9R97

Thanks in advance.

Nic-Ma commented:
Hi @wyli ,

BTW, I already added the doc-string in `CacheDataset` to highlight that it requires `persistent_workers=False` in DataLoader:
https://github.com/Project-MONAI/MONAI/pull/2436/files#diff-e682e654b07b9d753cac73f5a782e70644c82a84d30ebc757ec294a261845108R541

Thanks.

ericspod commented on monai/data/dataset.py:
```diff
@@ -130,6 +130,9 @@ class PersistentDataset(Dataset):
     Subsequent uses of a dataset directly read pre-processed results from `cache_dir`
     followed by applying the random dependant parts of transform processing.
 
+    If want to update the input data and recompute cache content during training, just call `update_data()`
```
Might be easier to say "During training call `update_data()` to update input data and recompute cache content.

Nic-Ma replied:
sure, updated in the latest commit.

Thanks.

ericspod commented on monai/data/dataset.py:
```diff
@@ -458,6 +479,7 @@ def _cachecheck(self, item_transformed):
         if self._read_env is None:
             self._read_env = self._fill_cache_start_reader()
         with self._read_env.begin(write=False) as txn:
+            print("!!!!!!!!", item_transformed)
```
Print statement shouldn't be here?

Nic-Ma replied:
sorry for the typo, removed it in the latest commit.

Thanks.

ericspod commented on monai/data/dataset.py:
```diff
@@ -515,6 +537,9 @@ class CacheDataset(Dataset):
     ``RandCropByPosNegLabeld`` and ``ToTensord``, as ``RandCropByPosNegLabeld`` is a randomized transform
     and the outcome not cached.
 
+    If want to update the input data and recompute cache content during training, just call `update_data()`
```
Same with this sentence and following ones, it can be condensed down a bit.

Nic-Ma replied:
Sure, updated all related in the latest commit.

Thanks.

wyli reviewed (approved):
thanks, just need to change the docstring `update_data()` to `set_data()`
